lib/ExtraCorres.thy

```diff
@@ -238,4 +238,224 @@ lemma hoare_from_abs_inv:
+    apply (insert cond)
+    apply clarsimp
+   apply (prop_tac "corres_underlying srel False nf' rrel (P and C r) (P' and C' r') (B r) (B' r')")
+    apply simp
+   apply (fastforce simp: corres_underlying_def)
+  apply clarsimp
+  apply (drule_tac x="(rv,t)" in bspec)
+   apply assumption
+  apply (prop_tac "P' conc_s")
+   apply (insert body_inv)
+   apply (fastforce simp: valid_def)
+  apply (fastforce intro: body_inv simp: valid_def)
+  done
+
+lemma whileLoop_terminates_cross:
+  assumes body_corres:
+           "\<forall>r r'. rrel r r'
+                    \<longrightarrow> corres_underlying srel False nf' rrel (P and C r) (P' and C' r')
+                                                              (B r) (B' r')"
+  assumes cond: "\<And>r r' s s'. \<lbrakk>rrel r r'; (s, s') \<in> srel; P s; P' s'\<rbrakk> \<Longrightarrow> C r s = C' r' s'"
+  assumes body_inv: "\<And>r. \<lbrace>P and C r\<rbrace> B r \<lbrace>\<lambda>_. P\<rbrace>"
+                    "\<And>r'. \<lbrace>P' and C' r'\<rbrace> B' r' \<lbrace>\<lambda>_. P'\<rbrace>"
+  assumes abs_termination: "\<And>r s. P s \<Longrightarrow> whileLoop_terminates C B r s"
+  assumes rel: "\<exists>r s. rrel r r' \<and> (s,s') \<in> srel \<and> P s"
+  assumes P': "P' s'"
+  shows "whileLoop_terminates C' B' r' s'"
+  apply (insert assms)
+  apply (fastforce dest!: whileLoop_terminates_cross_helper)
+  done
+
+lemma whileLoop_no_fail_abs:
+  assumes cond:
+          "\<And>r r' s s'. \<lbrakk>rrel r r'; (s, s') \<in> srel; P s; P' s'\<rbrakk> \<Longrightarrow> C r s = C' r' s'"
+  assumes body_corres:
+           "\<forall>r r'. rrel r r'
+                    \<longrightarrow> corres_underlying srel False True rrel (P and C r) (P' and C' r')
+                                                              (B r) (B' r')"
+  assumes body_inv: "\<And>r. \<lbrace>P and C r\<rbrace> B r \<lbrace>\<lambda>_. P\<rbrace>"
+                    "\<And>r'. \<lbrace>P' and C' r'\<rbrace> B' r' \<lbrace>\<lambda>_. P'\<rbrace>"
+  assumes nf: "\<And>r. no_fail (P and C r) (B r)"
+  assumes abs_termination: "\<And>r s. P s \<Longrightarrow> whileLoop_terminates C B r s"
+  assumes rel: "\<exists>r s. rrel r r' \<and> (s,s') \<in> srel \<and> P s"
+  assumes rrel: "\<forall>r'. \<exists>r. rrel r r'"
+  assumes P': "P' s'"
+  shows "\<not> snd (whileLoop C' B' r' s')"
+  apply (rule_tac I="\<lambda>rv' s'. \<exists>rv s. (s, s') \<in> srel \<and> rrel rv rv' \<and> P s \<and> P' s'"
+              and R="{((r', s'), r, s). C' r s \<and> (r', s') \<in> fst (B' r s)
+                                        \<and> whileLoop_terminates C' B' r s}"
+               in not_snd_whileLoop)
+    apply (insert assms)[1]
+    apply blast
+   apply (rename_tac conc_r s)
+   apply (clarsimp simp: validNF_def)
+   apply (intro conjI)
+    apply (intro hoare_vcg_conj_lift_pre_fix; (solves wpsimp)?)
+      apply (prop_tac "\<exists>abs_r. rrel abs_r conc_r")
+       apply (insert assms)[1]
```

Comment:
insert with [1] shouldn't need the [1] as insert does not affect other subgoals... I claim these should all have the [1] removed in this lemma and any other lemmas that use this pattern

Reply:
`insert` does affect all subgoals

Reply:
for real? that's crazy! I take that comment back then, but it's still a proof smell I'd rather not keep around :/

Reply:
I'm sure those old proofs of mine could be substantially improved. But one other reason for restricting things with an `(insert blah)[1]` (apart from `fastforce` instantiation things incorrectly) is that with many assumptions, it really clogs the proof state. These low-level proofs are sometimes not that pleasant and it seems nice to have at least a few points in the proof where someone can put their cursor and see a proof state that makes *some* sense

Reply:
Yes, I'd usually go for `using assms` these days, but that doesn't work if it's followed by `drule`, because it wants to do rule chaining. No real nice option for this specific case that I can see.

Reply:
In this particular proof, there's lots of `apply (insert assms)[1]` followed by a method that solves the subgoal. In most cases, you can either give assms to the method (simp/fastforce/etc) or use `using`. Maybe even something like `subgoal using assms by fastforce` or `subgoal using assms by - blast` ... I'm not seeing the case with the assms followed by drule in this proof at least.